internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -553,6 +636,103 @@ func (r *BMACReconciler) findBMH(ctx context.Context, agent *adiiov1alpha1.Agent
+	if result, err := controllerutil.CreateOrUpdate(ctx, spokeClient, bmhSpoke, mutateFn); err != nil {
+		r.Log.WithError(err).Errorf("failed to ensure spoke BareMetalHost %s/%s", bmhSpoke.Namespace, bmhSpoke.Name)
+		return nil, err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Spoke BareMetalHost created")
+	}
+	return bmhSpoke, nil
+}
+
+func (r *BMACReconciler) ensureSpokeMachine(ctx context.Context, spokeClient client.Client, bmh *bmh_v1alpha1.BareMetalHost, clusterDeployment *hivev1.ClusterDeployment) (*machinev1beta1.Machine, error) {
+	machineSpoke, mutateFn := r.newSpokeMachine(bmh, clusterDeployment)
+	if result, err := controllerutil.CreateOrUpdate(ctx, spokeClient, machineSpoke, mutateFn); err != nil {
+		r.Log.WithError(err).Errorf("failed to ensure spoke Machine %s/%s", machineSpoke.Namespace, machineSpoke.Name)
+		return nil, err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Spoke Machine created")
+	}
+	return machineSpoke, nil
+}
+
+func (r *BMACReconciler) ensureSpokeBMHSecret(ctx context.Context, spokeClient client.Client, bmh *bmh_v1alpha1.BareMetalHost) (*corev1.Secret, error) {
+	secretName := bmh.Spec.BMC.CredentialsName
+	secret := &corev1.Secret{}
+	err := r.Get(ctx, types.NamespacedName{Namespace: bmh.Namespace, Name: secretName}, secret)
+	if err != nil {
+		return secret, err
+	}
+	secretSpoke, mutateFn := r.newSpokeBMHSecret(secret, bmh)
+	if result, err := controllerutil.CreateOrUpdate(ctx, spokeClient, secretSpoke, mutateFn); err != nil {
+		r.Log.WithError(err).Errorf("failed to ensure spoke BareMetalHost Secret %s/%s", secretSpoke.Namespace, secretSpoke.Name)
+		return nil, err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Spoke BareMetalHost created")
+	}
+	return secretSpoke, nil
+}
+
+func (r *BMACReconciler) newSpokeBMH(bmh *bmh_v1alpha1.BareMetalHost, machine *machinev1beta1.Machine) (*bmh_v1alpha1.BareMetalHost, controllerutil.MutateFn) {
+	bmhSpoke := &bmh_v1alpha1.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      bmh.Name,
+			Namespace: bmh.Namespace,
+		},
+	}
+	mutateFn := func() error {
+		bmhSpoke.Spec = bmh.Spec
+		// The host is created by the baremetal operator on hub cluster. So BMH on the spoke cluster needs
+		// to be set to externally provisioned
+		bmhSpoke.Spec.ExternallyProvisioned = true
+		// If the Image field is filled in, ExternallyProvisioned is ignored. So remove the Image field from spec
+		bmhSpoke.Spec.Image = &bmh_v1alpha1.Image{}
+		bmhSpoke.Spec.ConsumerRef.Name = machine.Name
+		bmhSpoke.Spec.ConsumerRef.Namespace = machine.Namespace
+		// copy annotations. hardwaredetails annotations is needed for automatic csr approval
+		bmhSpoke.ObjectMeta.Annotations = bmh.ObjectMeta.Annotations
+		return nil
+	}
+
+	return bmhSpoke, mutateFn
+}
+
+func (r *BMACReconciler) newSpokeBMHSecret(secret *corev1.Secret, bmh *bmh_v1alpha1.BareMetalHost) (*corev1.Secret, controllerutil.MutateFn) {
```

Comment:
The `newSpokeSecret` and the `newSpokeMachine` function don't seem to be doing much, tbh. I think we could probably merge them into the caller function.